src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisResult.cs

```diff
@@ -13,195 +14,148 @@ namespace Microsoft.CodeAnalysis.Diagnostics
         {
-            Debug.Assert(getLocalDiagnostics || getNonLocalDiagnostics);
-
-            var builder = ImmutableArray.CreateBuilder<Diagnostic>();
-            if (getLocalDiagnostics)
+            if (SyntaxDiagnostics.Count > 0 || SemanticDiagnostics.Count > 0 || CompilationDiagnostics.Count > 0)
             {
-                if (!analysisScope.IsTreeAnalysis)
+                var builder = ImmutableArray.CreateBuilder<Diagnostic>();
+                if (getLocalSyntaxDiagnostics)
                 {
-                    AddAllLocalDiagnostics_NoLock(_localSyntaxDiagnosticsOpt, analysisScope, builder);
-                    AddAllLocalDiagnostics_NoLock(_localSemanticDiagnosticsOpt, analysisScope, builder);
+                    AddLocalDiagnostics(SyntaxDiagnostics, excludedAnalyzers, builder);
                 }
-                else if (analysisScope.IsSyntaxOnlyTreeAnalysis)
+
+                if (getLocalSemanticDiagnostics)
                 {
-                    AddLocalDiagnosticsForPartialAnalysis_NoLock(_localSyntaxDiagnosticsOpt, analysisScope, builder);
+                    AddLocalDiagnostics(SemanticDiagnostics, excludedAnalyzers, builder);
                 }
-                else
+
+                if (getNonLocalDiagnostics)
                 {
-                    AddLocalDiagnosticsForPartialAnalysis_NoLock(_localSemanticDiagnosticsOpt, analysisScope, builder);
+                    AddNonLocalDiagnostics(CompilationDiagnostics, excludedAnalyzers, builder);
                 }
-            }
 
-            if (getNonLocalDiagnostics && _nonLocalDiagnosticsOpt != null)
-            {
-                AddDiagnostics_NoLock(_nonLocalDiagnosticsOpt, analysisScope, builder);
+                return builder.ToImmutable();
             }
 
-            return builder.ToImmutableArray();
+            return ImmutableArray<Diagnostic>.Empty;
         }
 
-        private static void AddAllLocalDiagnostics_NoLock(
-            Dictionary<SyntaxTree, Dictionary<DiagnosticAnalyzer, List<Diagnostic>>> localDiagnostics,
-            AnalysisScope analysisScope,
+        private static void AddLocalDiagnostics(
+            ImmutableDictionary<SyntaxTree, ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<Diagnostic>>>  localDiagnostics,
+            ImmutableHashSet<DiagnosticAnalyzer> excludedAnalyzers,
             ImmutableArray<Diagnostic>.Builder builder)
         {
-            if (localDiagnostics != null)
+            foreach (var diagnosticsByTree in localDiagnostics)
             {
-                foreach (var localDiagsByTree in localDiagnostics.Values)
+                foreach (var diagnosticsByAnalyzer in diagnosticsByTree.Value)
                 {
-                    AddDiagnostics_NoLock(localDiagsByTree, analysisScope, builder);
-                }
-            }
-        }
+                    if (excludedAnalyzers.Contains(diagnosticsByAnalyzer.Key))
```

Comment:
I expect most common usage to be GetDiagnostics(allAnalyzers), so using excludedAnalyzers saves allocations. For per-analyzer diagnostics, one would normally directly use maps being exposed here.